proposals/p3646.md

```diff
@@ -0,0 +1,252 @@
+    -   Consistent use of tuple field indexes can be used to support code that
+        adds new tuple elements over time.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   This feature allows tuple access to be written more concisely than
+        pattern matching would allow.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   This feature provides a migration syntax for existing use of `.first`,
+        `.second`, and `std::get<I>`. The permission to use expressions rather
+        than only literals supports migration of `std::get<expression>`.
+
+Principles:
+
+-   [Low context sensitivity](/docs/project/principles/low_context_sensitivity.md).
+    -   We look only at the character immediately before a numeric literal to
+        determine whether it is lexed as a tuple index that stops before the
+        next `.` or as a general numeric literal.
+
+## Alternatives considered
+
+### Alternative lexing rule
+
+We could lex `.0`, `.1`, ... as a single token rather than as separate `.` and
+`0`, `1`, ... tokens. This would somewhat simplify the lexing rules, because
+they would no longer be contextual. We choose to not do this because:
+
+-   This would be inconsistent with our handling of `struct.fieldname`.
+-   Either `tuple . 0` would be invalid, unlike `struct.fieldname`, or it would
+    need to use a distinct grammar production from `tuple.0`.
+
+We could lex an integer literal when the previous token is `.`, regardless of
+whether the literal follows the `.` immediately. For example, we could treat
+
+```carbon
+let n: i32 = ((1, 2, 3), 4) . 0.1;
+```
+
+as tuple indexing, rather than as a tuple followed by a `.` and a real literal.
+We choose to not do this because:
+
+-   The `0.1` literal in this case looks like a real literal, not tuple
+    indexing, so this would likely cause surprise for readers.
+-   This would make the context-sensitive lexing be non-local. The chosen rule
+    can be interpreted as lexing `.[0-9]*` as a single lexeme, but forming two
+    tokens from it, whereas this alternative rule would be much more firmly a
+    context-sensitive lexing rule.
+-   This would be inconsistent with the Rust behavior, which this feature
+    otherwise matches.
+
+### Decimal indexing restriction
+
+Rust restricts tuple indexes to being decimal integers:
+
+```rust
+// OK
+(1, 2, 3).0;
+
+// Error, no field named `0x0`.
+(1, 2, 3).0x0;
+```
+
+We could apply the same restriction in Carbon. However, this restriction would
+introduce inconsistency between `.0x0` and `.(0x0)`, the latter of which does
+not exist in Rust. While we don't imagine a lot of use for these indexes, it's
+possible that a binary index might be useful in some special cases, and there
+doesn't seem to be a reason to disallow this beyond coding style.
```

Comment:
FWIW, I lean towards the Rust restriction.

There are already lexical inconsistencies here, for example the fact that it doesn't form a real literal when followed by a `.`.

Personally, I feel like it will be easier to teach that these are "positional names" with their own, much narrower set of rules than integer literals.

But interested in what others think about this specific issue -- and I wonder if it would be worth extract this into a specific issue for leads to record this aspect of the decision?

Reply:
I don't feel strongly about this at all. If we want to say that we treat `.0`, `.1`, ... as the *names* of the elements of a tuple, I think that's a reasonable reason to add a restriction. If we do add such a restriction, I think we should also forbid digit separators in these indexes. (I can't really imagine ever wanting digit separators here -- indexing a `1_000`-tuple seems ... implausible, and allowing them would not be compatible with the idea that these are numeric field names.)

Reply:
FWIW, you'd still be able to use digit separators if useful: `huge_tuple.(1_000)`.

Reply:
Done.